### PR TITLE
[SPARK-20204][SQL][Followup] SQLConf should react to change in default timezone settings

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Make sure SESSION_LOCAL_TIMEZONE reflects the change in JVM's default timezone setting. Currently several timezone related tests fail as the change to default timezone is not picked up by SQLConf.

## How was this patch tested?
Added an unit test in ConfigEntrySuite

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala [hunk 1]
```diff
@@ -147,6 +147,15 @@ private[spark] class TypedConfigBuilder[T](
     }
   }
 
+  /** Creates a [[ConfigEntry]] with a function has a default value */
+  def createWithDefaultFunction(defaultFunc: () => T): ConfigEntry[T] = {
+    val entry =
+      new ConfigEntryWithDefaultFunction[T](parent.key, defaultFunc, converter,
+      stringConverter, parent._doc, parent._public)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala]: nit
```
val entry = new XXX(
  param1, param2, ...)
```

**Comment by cloud-fan** [comment]
LGTM, we can have a unit test to make sure the default value can change after the conf entry is created, in `ConfigEntrySuite`

**Comment by dilipbiswal** [comment]
@cloud-fan Sure. I will add the unit test.

**Comment by viirya** [comment]
LGTM for current change. Look forward for the test to add.

File: core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala [hunk 2]
```diff
@@ -51,6 +52,26 @@ class ConfigEntrySuite extends SparkFunSuite {
     assert(conf.get(dConf) === 20.0)
   }
 
+  test("conf entry: timezone") {
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala]: we are testing `ConfigEntryWithDefaultFunction`, not timezone

File: core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala [hunk 3]
```diff
@@ -51,6 +52,26 @@ class ConfigEntrySuite extends SparkFunSuite {
...
+    val tzStart = TimeZone.getDefault().getID()
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala]: The test can be as simple as:
```
var data = 0
val conf = new SparkConf()
val iConf = ConfigBuilder(testKey("int")).intConf.createWithDefaultFunction(() => data)
assert(conf.get(iConf) === 0)
data = 2
assert(conf.get(iConf) === 2)
```

File: core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala (hunk 2 above)
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala]: also move this test to the end, it's weird to put this test in the middle of different config types tests.

File: core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala (hunk 3 above)
Comment (dilipbiswal) [inline core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala]: @cloud-fan Thank you :-). Its so much simpler. I somehow was trying to use the exact same function that we use in the code.

File: core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala [hunk 4]
```diff
@@ -147,6 +147,14 @@ private[spark] class TypedConfigBuilder[T](
     }
   }
 
+  /** Creates a [[ConfigEntry]] with a function has a default value */
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala]: `Creates a [[ConfigEntry]] with a function has a default value`
->
`Creates a [[ConfigEntry]] with a function to determine the default value`

File: core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala [hunk 5]
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.internal.config
 
+import java.util.TimeZone
```
Comment (gatorsmile) [inline core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala]: Remove it?

File: core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala [hunk 6]
```diff
@@ -51,6 +52,8 @@ class ConfigEntrySuite extends SparkFunSuite {
     assert(conf.get(dConf) === 20.0)
   }
 
+
+
```
Comment (gatorsmile) [inline core/src/test/scala/org/apache/spark/internal/config/ConfigEntrySuite.scala]: Nit: two more extra spaces

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -752,7 +752,7 @@ object SQLConf {
     buildConf("spark.sql.session.timeZone")
       .doc("""The ID of session local timezone, e.g. "GMT", "America/Los_Angeles", etc.""")
       .stringConf
-      .createWithDefault(TimeZone.getDefault().getID())
+      .createWithDefaultFunction(() => TimeZone.getDefault().getID())
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `TimeZone.getDefault().getID()`
->
`TimeZone.getDefault.getID`

**Comment by gatorsmile** [comment]
LGTM with minor comments.

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

**Comment by viirya** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dilipbiswal** [comment]
@viirya @cloud-fan @gatorsmile Thanks a lot.

**Pull request closed without merging**